Purchases/Caching/RCDeviceCache.m

```diff
@@ -284,35 +287,77 @@ - (void)deleteAttributesIfSyncedForAppUserID:(NSString *)appUserID {
         NSMutableDictionary <NSString *, NSDictionary *>
             *groupedAttributes = self.storedAttributesForAllUsers.mutableCopy;
         [groupedAttributes removeObjectForKey:appUserID];
-        [self.userDefaults setObject:groupedAttributes forKey:RCSubscriberAttributesKey];
+        [self storeAttributesForAllUsers:groupedAttributes];
     }
 }
 
 # pragma mark - subscriber attributes migration from per-user key to grouped key
 
-- (void)migrateSubscriberAttributesIfNeededForAppUserID:(nonnull NSString *)appUserID {
+- (void)cleanupSubscriberAttributes {
     @synchronized (self) {
-        NSDictionary *legacySubscriberAttributes = [self valueForLegacySubscriberAttributes:appUserID];
-        if (legacySubscriberAttributes != nil) {
-            [self migrateSubscriberAttributes:legacySubscriberAttributes withAppUserID:appUserID];
-        }
+        [self migrateSubscriberAttributes];
+        [self deleteSyncedSubscriberAttributesForOtherUsers];
     }
 }
 
-- (void)migrateSubscriberAttributes:(nonnull NSDictionary *)legacyAttributes
-                      withAppUserID:(nonnull NSString *)appUserID {
-    NSMutableDictionary *allSubscriberAttributes = ([self.userDefaults dictionaryForKey:RCSubscriberAttributesKey]
-                                                    ?: @{}).mutableCopy;
-    NSMutableDictionary *currentAttributesForAppUserID = allSubscriberAttributes[appUserID]
-                                                         ?: @{}.mutableCopy;
-    NSMutableDictionary *mutableLegacyAttributes = legacyAttributes.mutableCopy;
-    [mutableLegacyAttributes addEntriesFromDictionary:currentAttributesForAppUserID];
+- (void)migrateSubscriberAttributes {
+    NSArray *appUserIDsWithLegacyAttributes = [self appUserIDsWithLegacyAttributes];
+    NSMutableDictionary *attributesInNewFormat = self.storedAttributesForAllUsers.mutableCopy;
+    for (NSString *appUserID in appUserIDsWithLegacyAttributes) {
+        NSDictionary *legacyAttributes = ([self valueForLegacySubscriberAttributes:appUserID] ?: @{}).mutableCopy;
```

Comment:
If this an `NSDictionary`, the `mutableCopy` is not needed

Reply:
good catch, I refactored and forgot to remove the mutableCopy part